### PR TITLE
MIDI for light:  add settings ui

### PR DESCRIPTION
**Pull request by ekordas**
Adds Setting definition for UI of the Midi for Light Script

### CODE REVIEW

**Comment by ronso0** [comment]
Thank you for this PR.
Please sign the [Mixxx Contributor Agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ) and comment here when you have done so. It gives us permission to distribute your contribution under the GPL v2 or later license and the Apple Mac App Store. It is also helpful for us to have contact information for contributors in case we may need it in the future.

Do you mind attaching a screenshot of the new settings?

**Comment by ekordas** [comment]
I allready signed the agreement.

Here is a screenshot:
midi_for_light-settings

**Comment by ronso0** [comment]
Okay, thank you, I didn't remember you already worked on MIDI for lights.

The VU checkboxes could use more human-readable labels, not just the internal variable names, no?

I didn't work with the controller settings, yet, wasn't there an option/flag that would arange group items horizontally if the preference page allows it?

**Comment by ekordas** [comment]
I didnt work with the controller settings either, If there is a documentation for it i can set other flags. Leaving the Orientation to unset flipps checkboxes to a row while resizing the settings wich is really destroying the layout:
Bildschirmfoto vom 2024-10-07 00-00-49
So i leave it to vertical.
The Labels are changed in the next commit.

**Comment by JoergAtGithub** [comment]
> If there is a documentation for it i can set other flags.

@acolombier Do we have this described in the manual?

The best reference is the mapping for the S4 Mk3: https://github.com/mixxxdj/mixxx/pull/12995

**Comment by ekordas** [comment]
The Mapping xml file is described at https://github.com/mixxxdj/mixxx/wiki/MIDI%20controller%20mapping%20file%20format but not the ui Stuff. I Used the diff mentioned as reference.

**Comment by ekordas** [comment]
The <row orientation="vertical"> seems to do the intended behavior, but if the vertical list of checkboxes is flipped to a row  it does not flip back. And it does not matter how slim you resize the window. 
The section above does change to a row if the Size allows it and back, but it seems there is a Problem with a long list of checkboxes.
Since the UI is more aimed at developers and technical applications I personally would not bother making it pretty to much. It only requires a little scrolling.
Following is a new attempt:
Bildschirmfoto vom 2024-10-07 00-52-49

**Comment by ekordas** [comment]
Is anything missing or is everyone OK with the design?

**Comment by ronso0** [comment]
I'm not familiar with controller settings.
@JoergAtGithub wdyt?

**Comment by JoergAtGithub** [comment]
The settings syntax looks goot to me. But I'm not familar with MIDI for Light.

File: res/controllers/Midi_for_light-scripts.js [hunk 1]
```diff
@@ -16,33 +16,33 @@ function midi_for_light() {}
...
-var enable_vu_right_current = false; // set to false if you not need VU right current
-var enable_vu_right_average_min = false; // set to false if you not need VU right average min
-var enable_vu_right_average_mid = false; // set to false if you not need VU right average mid
-var enable_vu_right_average_max = false; // set to false if you not need VU right average max
-var enable_vu_right_average_fit = true; // set to false if you not need VU right average fit
-var enable_vu_right_current_meter = false; // set to false if you not need VU right current meter
-var enable_vu_right_average_meter = false; // set to false if you not need VU right average meter
-var deck_ending_time = 15; // set a time (in seconds) in which the playing track is considered to be ending
-var deck_ending_priority_factor = 0.9; // decrease the priority of the ending track by this factor
+var midi_channel = engine.getSetting("midi_channel"); // set midi_channel. Valid range: 1 to 16.
```
Comment (fwcd) [inline res/controllers/Midi_for_light-scripts.js:19]: In previous mappings we've "defensively" added the default values here too (either using the nullish coalescing operator `??` or `||`). I don't think that should be necessary, so this looks good to me, but just to confirm, is this safe @acolombier?

Comment (JoergAtGithub) [reply res/controllers/Midi_for_light-scripts.js:19]: AFAIK, the default values there were only necessarry for backward compatibility of the mapping to 2.4 without the controller settings feature.

Comment (fwcd) [reply res/controllers/Midi_for_light-scripts.js:19]: Ah interesting, so `engine.getSetting` was already available in 2.4, but invocations always returned `undefined`?

Comment (JoergAtGithub) [reply res/controllers/Midi_for_light-scripts.js:19]: It does not exist in 2.4, and is therefore `undefined`.

Comment (fwcd) [reply res/controllers/Midi_for_light-scripts.js:19]: You can't call `undefined` without getting a `TypeError` though.

Comment (ekordas) [reply res/controllers/Midi_for_light-scripts.js:19]: I can confirm the mapping does not work with the proposed fallback. Changing the line to
`var midi_channel = engine.getSetting("midi_channel") || 1;`
causes an error in Version 2.4:
`Uncaught exception: file:///C:/Program Files/Mixxx/controllers/Midi_for_light-scripts.js:19: TypeError: Property 'getSetting' of object ControllerScriptInterfaceLegacy(0x16b9d336520) is not a function`

Comment (ekordas) [reply res/controllers/Midi_for_light-scripts.js:19]: `var midi_channel = engine.getSetting ? engine.getSetting("midi_channel") : 1;` seems to work

Comment (fwcd) [reply res/controllers/Midi_for_light-scripts.js:19]: Depending on how new the JS engine is you might be able to do

```ts
engine.getSetting?.("midi_channel") ?? 1
```

But tbh if we're merging this to 2.5, it shouldn't be needed anyway (and I don't think there'll be many 2.4 releases anymore apart from minor bug fixes)

**Comment by acolombier** [comment]
I have added some [documentation in the wiki](https://github.com/mixxxdj/mixxx/wiki/MIDI-Scripting#settings-api), hopefully this can help refining the UI a bit.

**Comment by ekordas** [comment]
So, should I
a) implement a fallback and re target to 2.4 (previous underlying patch was considered to go into 2.5)
b) leave target to 2.4
and
1) retouch the ui? Personally i am quite happy now...
2) you have some style issues to fix?
3) leave as is?

**Comment by fwcd** [comment]
I'd say targeting 2.5 is fine, along with keeping everything as is. From a quick glance the PR looks good, I'll see if I find some time to test it soon.

**Comment by fwcd** [comment]
I have tested this in QLC+[^1] and it seems to work well. Thanks!

[^1]: Just in case anyone else wants to experiment with this, here's a workspace that maps every MIDI control (it's a bit fiddly to do that manually): [Mixxx MIDI.qxw.zip](https://github.com/user-attachments/files/18055597/Mixxx.MIDI.qxw.zip) This is configured to use a device named "Loop", which is a small virtual MIDI device created by another small tool I wrote: https://github.com/fwcd/midiloop

**Pull request merged**